.changeset/tall-waves-impress.md

```diff
@@ -0,0 +1,58 @@
+---
+'astro': minor
+---
+
+Change Action cookie redirects to an opt-in feature users can implement using middleware. This means Actions submitted from an HTML form action will no longer redirect to the destination as a GET request. Instead, the page will be rendered as a POST result.
+
+This is meant to address the 4 KB size limit users have encountered when calling actions from an HTML form action. It can be difficult to predict the size of an action result for large validation errors or longer return values, like AI chatbot results.
+
+So, we've introduced a new `getActionContext()` utility to let you decide how action results are handled from middleware. You may choose to implement the existing cookie redirect from Astro v4, or implement forwarding with your own session storage.
```

Comment:
```suggestion
Changes the default redirect behavior of actions submitted from an HTML form and adds a new `getActionContext()` utility to Astro Actions to let you decide how action results are handled from middleware.

Actions submitted from an HTML form action will no longer redirect to the destination as a GET request. Instead, the page will be rendered as a POST result. Cookie redirects can still be implemented in middleware, but are no longer the default. `getActionContext()` additionally allows you to implement forwarding with your own session storage:

// example here of what a "new", "normal" action would look like
// don't show using cookies, since you'll have that below

This change is meant to address the 4 KB size limit users have encountered when calling actions from an HTML form action. It can be difficult to predict the size of an action result for large validation errors or longer return values, like AI chatbot results. `getActionContext()` [DOES / ALLOWS YOU TO... feels like we just want a super quick sentence here that emphasizes or gives some examples of what this unlocks.... (we've said "let you decide" but maybe something specific here wraps this up nicely]
```
What would you think about a structure like this? (mostly just reorganized what you had) I will totally admit that the 2nd paragraph is a bit clunky, and you'd know a nicer way to make those ideas fit/flow. I also admit that this de-emphasizes cookies redirects (less prominent placement) in favour of a "now you specify how you want things handled", which *is* a choice, but maybe *feels* like what a changelog should be doing? Just for your consideration here!

In particular, I think it would be nice to show an example of doing this the new way, without cookies. And, I think that introduces `getActionContext()` way earlier, and doesn't make people read your Cool Story Bro of what all was happening, and why you decided to introduce this new function which doesn't show up until the very end.

Reply:
Also want to say that users probably won't need to decide how actions are handled for 99% of use cases. The browser default is fine. I'd describe middleware usage as advanced customization that people can take advantage of when browser defaults won't work for them